### PR TITLE
feat(mcp): stdio transport over a byte-level stdin read effect (i66E)

### PR DESCRIPTION
**Pull request by sergey-shandar**
Implements the stdio transport from `issues/66E-mcp-stdio-transport.md` (now `Status: done`), unblocking `i66E-mcp-cas-server`.

## What

- **`fs/effects/node` — byte-level `read` effect.** A new `Read = ['read', (stream) => number | null]`, the dual of `write`: one byte (`0`–`255`) or `null` at EOF. `readLine` is a pure combinator over it — it accumulates bytes into a cons-list (O(1)/byte) and reverses + UTF-8-decodes once at the newline/EOF terminator (O(n)). Reading one byte at a time means a line never over-reads past its terminator, so `readLine` is self-contained with no cross-call buffer. The virtual runner serves a byte stream; the Node interpreter reads `process.stdin` one byte at a time, distinguishing EOF from "not yet readable" via a `readable`/`end` wait that detaches both listeners after each read.
- **`fs/mcp/stdio` — `stdioTransport(step)`.** Wraps an `mcpStep`-shaped `(value) => Effect<O, Response | null>` in the read→parse→dispatch→write loop, as a recursive effect that stays in the pure effect model and is fully testable against a mock stdin/stdout with no real process. Edge cases per the issue: EOF → clean shutdown; notification (`null` step result) → no reply; malformed JSON line → JSON-RPC `-32700` parse-error response. Responses serialize with `undefined`-valued fields omitted, matching `JSON.stringify`.
- **`fs/json/parser` — strict trailing commas.** The transport parses untrusted input, and the repo parser previously accepted trailing commas (`[1,]`, `{"a":1,}`). Made it require a value after `[`/`,`/`:` and a string key after a `,` in an object, so it matches strict `JSON.parse` (empirically the only divergence — every JS-only number/quote/keyword form is already rejected by the tokenizer). Its two trailing-comma proofs moved from `valid` to `invalid`.

## Tests

Proofs drive `stdioTransport` against a mock stdin byte stream and assert the bytes written to a mock stdout: EOF, request→response, notification→silence, malformed→`-32700`, trailing-comma→`-32700`, `undefined`-field omission, an unterminated final line, and a mixed multi-line session. `fs/mcp/stdio/module.f.ts` and `fs/effects/node/module.f.ts` are at 100% line/branch/function coverage; full suite green.

## Out of scope

The `cas mcp` CLI wiring is the follow-on `i66E-fjs-cas-mcp-subcommand`; the generic `casMcpHandlers` adapter is `i66E-mcp-cas-server`. This PR owns only the transport primitive.

### CODE REVIEW

**Comment by sergey-shandar** [comment]
@codex review

File: fs/mcp/stdio/module.f.ts [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    <O extends Operation>(step: Step<O>): Effect<ReadLine | Write | O, void> =>
+    readline().step(line =>
+        line === null
+            ? pure(undefined)
+            : handleLine(step)(line))
+
+const handleLine =
+    <O extends Operation>(step: Step<O>) =>
+    (line: string): Effect<ReadLine | Write | O, void> => {
+        const [t, value] = parse(tokenize(stringToList(line)))
```
[the remark this answers is not in the document]
Comment (sergey-shandar) [reply fs/mcp/stdio/module.f.ts:79]: Fixed in f30fcc0. Empirically the only divergence between the repo parser and strict `JSON.parse` was trailing commas — every JS-only form (`0x10`, `1_000`, `.5`, `01`, single quotes, `NaN`/`undefined`) is already rejected by the tokenizer — so I made `fs/json/parser` strict: a value is required after `[`/`,`/`:`, and a string key after a `,` in an object. The two trailing-comma cases moved from the parser's `valid` proofs to `invalid`, and a new transport proof asserts `{"jsonrpc":"2.0","method":"ping","id":1,}` now yields the `-32700` parse-error response instead of being dispatched.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by sergey-shandar** [comment]
@codex review

**Comment by sergey-shandar** [comment]
@codex review

File: fs/effects/node/module.f.ts [hunk 2]
```diff
@@ -222,6 +225,52 @@ export const log: Console = writeString('stdout')
...
+ * buffered; a final line lacking a trailing newline is returned in full.
+ */
+export const readLine = (stream: ReadConsoles): Effect<Read, string | null> => {
+    const loop = (acc: readonly number[]): Effect<Read, string | null> =>
+        read(stream).step(b =>
+            b === null
+                ? pure(acc.length === 0 ? null : utf8ListToString(acc))
+                : b === lf
+                    ? pure(utf8ListToString(acc))
+                    : loop([...acc, b]))
```
[the remark this answers is not in the document]
Comment (sergey-shandar) [reply fs/effects/node/module.f.ts]: Fixed in 5e36574. `readLine` now accumulates bytes into a cons-list by prepending (O(1)/byte) and reverses + UTF-8-decodes once at the newline/EOF terminator, so a large line is O(n) instead of O(n²). Coverage on `fs/effects/node/module.f.ts` stays at 100%.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: fs/effects/node/module.ts [hunk 3]
```diff
@@ -149,6 +149,34 @@ const writeAll = async (stream: NodeJS.WritableStream, data: Uint8Array): Promis
...
+        if (chunk !== null) {
+            return chunk[0]
+        }
+        if (stdin.readableEnded) {
+            return null
+        }
+        const ended = await Promise.race([
+            once(stdin, 'readable').then(() => false),
+            once(stdin, 'end').then(() => true),
+        ])
```
[the remark this answers is not in the document]
Comment (sergey-shandar) [reply fs/effects/node/module.ts]: Fixed in 5e36574. Replaced the `Promise.race` of `once(stdin,'readable')`/`once(stdin,'end')` with a `waitReadableOrEnd` helper that installs both listeners and removes both as soon as either fires, so an idle long-running server no longer accumulates stale `end`/`readable` listeners.

---
_Generated by [Claude Code](https://claude.ai/code)_

[the remark this answers is not in the document]
Comment (sergey-shandar) [reply fs/effects/node/module.ts]: Fixed in 5e36574 (same change as the related thread): `waitReadableOrEnd` now removes both the `readable` and `end` listeners as soon as either fires, so the losing listener is detached after each race and no `end` listeners pile up across idle gaps.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by sergey-shandar** [comment]
@codex review

File: issues/66E-mcp-stdio-transport.md [hunk 4]
```diff
@@ -70,14 +70,16 @@ transport. This issue owns only the transport primitive.
...
 
-- [ ] Add `readline` effect type to `fs/effects/node/module.f.ts` and its Node.js
-      interpreter (`process.stdin` / `readline` interface or async iterator).
-- [ ] Add `stdioTransport` (or equivalent combinator) that wraps a step function
-      in the read-parse-dispatch-write loop.
-- [ ] Proof / tests: drive `stdioTransport` against a mock stdin sequence (list of
-      lines) and assert the correct responses are written to a mock stdout — no
-      real process needed.
-- [ ] Handle edge cases: EOF mid-stream, malformed JSON lines, and `null` response
+- [x] Add a byte-level `read` effect (`Read`, the dual of `write`) to
```
[the remark this answers is not in the document]
Comment (sergey-shandar) [reply issues/66E-mcp-stdio-transport.md:73]: Done in 26dbb60 — set `Status: done` in `issues/66E-mcp-stdio-transport.md`, which unblocks the follow-on `i66E-mcp-cas-server`. (Per AGENTS.md, done issues are removed later in a separate cleanup pass, so I'm leaving the file in place for now.)

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by sergey-shandar** [comment]
@codex review

File: fs/mcp/stdio/module.f.ts [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+/** Serializes a JSON value with deterministic (sorted) key order. */
+const stringifyJson = stringify(sort)
+
+/** The parse-error response (`-32700`, `id: null`) for a malformed input line. */
+const parseErrorResponse: Response = { jsonrpc, error: parseError, id: null }
+
+/** Encodes a response as a newline-terminated UTF-8 line and writes it to `stdout`. */
+const writeResponse = (resp: Response): Effect<Write, void> =>
+    write('stdout', utf8(stringifyJson(resp as Unknown) + '\n'))
```
[the remark this answers is not in the document]
Comment (sergey-shandar) [reply fs/mcp/stdio/module.f.ts:57]: Fixed in ee9b1a7. `writeResponse`'s serializer now filters out `undefined`-valued entries at every object level via the `stringify` `mapEntries` hook (`e => sort(filter(p => p[1] !== undefined)(e))`), so an optional field like `error.data` or `nextCursor` left explicitly `undefined` is omitted — matching `JSON.stringify` — instead of throwing `TypeError` and aborting the loop. Added a proof that drives a step returning `{ result: { ok: true, nextCursor: undefined }, … }` and asserts the field is omitted. `fs/mcp/stdio/module.f.ts` stays at 100% coverage.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by sergey-shandar** [comment]
@codex review
